src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    public class ResponseCachingOptions
+    {
+        /// <summary>
+        /// Override default behavior for determining cacheability of an HTTP request.
+        /// </summary>
+        public Func<HttpContext, bool> RequestIsCacheableOverride { get; set; }
+
+        /// <summary>
+        /// Override default behavior for determining cacheability of an HTTP response.
+        /// </summary>
+        public Func<HttpContext, bool> ResponseIsCacheableOverride { get; set; }
+
+        /// <summary>
+        /// Create a custom segment that is appended to the default cache key.
+        /// </summary>
+        public Func<HttpContext, string> CreateCustomizedCacheKey { get; set; }
```

Comment:
We can consider passing the StringBuilder that contains the default key to the user. It may be more efficient if the user needs to concatenate many small segments since it saves creating a new instance of StringBuilder. But not sure how often this would arise and it may be a premature optimization. It would also have the side effect of the user being able to overwrite the default cache key.